### PR TITLE
Pyupgrade: Printf string formatting

### PR DESCRIPTION
**Pull request by colin99d**
A part of #827. This one is complicated so it will take me some more time. 

Please Note: I purposely did NOT implement the `--keep-percent-format` flag because it looks like the last flag I added was removed. If you want this flag in, just let me know.

Note: the issue of the /N{dog} format will plague us in UP030 and UP032. Whatever solution we find here needs to be applied in both of those places. Right now I am just thinking of looking for it, and if it exists, only showing a warning but not attempting top fix. I have NEVER seen this syntax in practice, and I think it is reasonable to give the people using it a warning, but don't fix.

### CODE REVIEW

**Comment by colin99d** [comment]
@charliermarsh I have hit a blocker and I would love some assistance. Pyupgrade uses the regex ["(?<!\\)(?:\\\\)*(\\N\{[^}]+\})"](https://github.com/asottile/pyupgrade/blob/main/pyupgrade/_string_helpers.py#L9) to get an outcome like below:
Screenshot 2023-01-13 at 8 41 24 PM

We cannot use this regex in rust, because the regex uses look behinds, which are [not implemented in rust](https://stackoverflow.com/questions/61485063/is-there-alternative-regex-syntax-to-avoid-the-error-look-around-including-loo). I tried taking the post's advice and reimplementing the regex without look behinds, "[^\\](?:\\\\)*(\\N\{[^}]+\})", but then the character before is included in the string, which causes the program to break:
Screenshot 2023-01-13 at 8 44 51 PM

After this I tried using the `fancy_regex` create; however, they [do not implement the `split` method](https://github.com/fancy-regex/fancy-regex/issues/104) which is required. With this I am unsure of what to do next. We could check for my ugly regex, and just ignore those (or report them as issues but not fix them). But that really bothers the perfectionist in me. Let me know how you think we should handle this.

**Comment by charliermarsh** [comment]
Why do we need / why does the regex exist? Is it to avoid treating named unicode characters as `{}`-style formatting placeholders?

**Comment by colin99d** [comment]
I know its for statements like these: `"%s \N{snowman}" % (a,)`, because they are currently causing issues in my code. However, I am not sure the specific reason they are needed.

**Comment by colin99d** [comment]
It looks like it exists so that `"%s \N{snowman}" % (a,)` does not turn into `"%s \N{{snowman}}" % (a,)`. If this happens we get the following error in python:
SyntaxError: (unicode error) 'unicodeescape' codec can't decode bytes in position 3-14: unknown Unicode character name

**Comment by charliermarsh** [comment]
Is there a way that we could run the regex, extract the match, then run a _second_ regex over the match to get a more precise range?

**Comment by colin99d** [comment]
> Is there a way that we could run the regex, extract the match, then run a _second_ regex over the match to get a more precise range?

Right now the regex I have developed that is the closest is `"[^\\](?:\\\\)*(\\N\{[^}]+\})"`. It has two issues:
1. It grabs one character before the match starts (this could in theory be fixed by your solution)
2. It would never match: `\N{snowman}" % (a,)`, I am not sure how to overcome this shortcoming

Additionally, I am not sure if we could use this "double regex" with methods like `split`. I have been trying to think of other ways we could do this. I am sure we could do something with tokens, but not sure how we could turn this back into a string simply.

**Comment by charliermarsh** [comment]
Ok, let me take a closer look. Don't worry about resolving it for now. I have to spend some time with the code to understand the problem before I can offer good advice :)

**Comment by colin99d** [comment]
Sounds good! I am working on the dict implementation, once I get that done this is solved besides the regex issue.

**Comment by colin99d** [comment]
@charliermarsh I am moving this out of draft now. This one was pretty complicated so I left as many notes as I could. Everything is working as I want it to besides the regex we discussed above. However, please not I departed from pyupgrade for two negative test cases (test cases where the linter should NOT run):
```
        # don't rewrite string-joins in dict literal
        '"%(ab)s" % {"a" "b": 1}',
```
Rustpython_ast automatically makes this string-join "ab". This means that not linting here would take more logic on our end, and would end up being less helpful for the user. If you can think of an edge case where this is an issue, let me know and I can fix.

```
        # don't rewrite strangely styled things
        '"%(a)s" % {"a"  :  1}',
```
The rust_python ast we are using doesnt have the spaces factored in, so they do not create issues. Fixing this one will require adding complicated logic, to decrease the user experience. As before, if you disagree just let me know.

**Comment by charliermarsh** [comment]
Thanks. I know I owe you a review here! Hasn’t slipped my mind. Just trying to find time since it’s a big PR and requires focus :)

**Comment by colin99d** [comment]
No worries! I figured it might be a day or two before you get this one! By far the most complicated one I have done.

**Comment by charliermarsh** [comment]
(Reviewing now.)

**Comment by charliermarsh** [comment]
I haven't finished reviewing the code yet, but I notice this is failing on a few cases:

```diff
@@ -244,8 +244,8 @@ def run(
     print(
-        "Writing merged info for %s slides (%s unrecognized)"
-        % (len(merged_slide_data), unrecognized_count)
+        "Writing merged info for {%s unrecognized:"}
+       .format(len(merged_slide_data), unrecognized_count)
     )
     merged_slide_data = sorted(
         merged_slide_data, key=lambda row: cast(str, row["Slides_slide_barcode"])
```

```py
     def composite_image(self, files: List[str]) -> List[str]:
         if len(files) != len(self.channels):
             raise ValueError(
-                "Expected one image (got %d) per channel (got %d)"
-                % (len(files), len(self.channels))
+                "Expected one image (got {got %d:"}
+               .format(len(files), len(self.channels))
             )

         composited_image = None
```

**Comment by colin99d** [comment]
I see you are making changes. Let me know once you are done and I can attack issues you find.

**Comment by charliermarsh** [comment]
I'll stop for now :)

**Comment by charliermarsh** [comment]
> I have NEVER seen this syntax in practice, and I think it is reasonable to give the people using it a warning, but don't fix.

FWIW, that's fine with me for now.

**Comment by charliermarsh** [comment]
> --keep-percent-format

Yeah we don't need to implement these flags IIUC because that's just equivalent to `--ignore UP031`.

File: src/rules/pyupgrade/helpers.rs [hunk 1]
```diff
@@ -0,0 +1,19 @@
+const KEYWORDS: [&str; 35] = [
+    "False", "None", "True", "and", "as", "assert", "async", "await", "break", "class", "continue",
+    "def", "del", "elif", "else", "except", "finally", "for", "from", "global", "if", "import",
+    "in", "is", "lambda", "nonlocal", "not", "or", "pass", "raise", "return", "try", "while",
+    "with", "yield",
+];
+
+/// This function currently does not work with python emoji synatx. For example "/N{dog}" will be
+/// converted to "/N{{dog}}" which is NOT correct and WILL cause issues. Please check for emoji
```
Comment (andersk) [inline src/rules/pyupgrade/helpers.rs:9]: It’s `\N{dog}`, not `/N{dog}`.

Comment (colin99d) [reply src/rules/pyupgrade/helpers.rs]: \N{facepalm}

**Comment by colin99d** [comment]
> 

Pyupgrade won't event attempt to fix these, time to find out why and get it fixed.

**Comment by colin99d** [comment]
@charliermarsh, unfortunately I was not able to fix the two issues you found tonight. However, I was able to determine that `parse_percent_format` is the function that is causing the issue. I was also able to run the same strings through the actual pyupgrade library, to determine the structs that should have been generated. This allowed me to create two failing unit tests. I am hoping to get this issue fixed by Sunday night at the latest. Also, I am wondering why pyupgrade does not change the strings at all, because the statement does not seem to violate any rules laid out in the unit tests.

**Comment by charliermarsh** [comment]
@colin99d - My suggestion is to try and use `CFormatString` from RustPython instead of regular expressions. You can see how it works in the f-string PR. But e.g.:

```rs
let Ok(format_string) = CFormatString::from_str(&left_string[1..left_string.len() - 1]) else {
    return;
};
```

This gives you structured data about the string that I _think_ we can use to reconstruct it. And the parser will be very robust. I can also take a crack at this if you're too tired of this one!

**Comment by colin99d** [comment]
> 

I have no problem doing it (always good to learn more tools for later PRs). Will try to have it done by this Sunday night.

**Comment by charliermarsh** [comment]
Ok cool, thank you -- as always, appreciate all the effort here! Let me know how it goes. I can't guarantee that it will work, since I haven't done the exercise myself, but I suspect it _should_, and would lead to a really strong solution.

**Comment by colin99d** [comment]
> Ok cool, thank you -- as always, appreciate all the effort here! Let me know how it goes. I can't guarantee that it will work, since I haven't done the exercise myself, but I suspect it _should_, and would lead to a really strong solution.

No problem! Also, would love to chat on Discord or something about next steps after I get pyupgrade done in the next couple weeks. Right now I am thinking about attacking pylint because that takes the most time in our CI and in pre-commit.

**Comment by charliermarsh** [comment]
@colin99d - Sounds great. I did setup a Discord, but... I haven't shared it with anyone yet, so it's just me :joy:

**Comment by colin99d** [comment]
> @colin99d - Sounds great. I did setup a Discord, but... I haven't shared it with anyone yet, so it's just me 😂

Awesome! Honestly a small private discord is probably the way to go, where you just invite people actively contributing. That way you dont get bots and spammers. Also, ruff should just work so once this tool is ready community members should not need a lot of support.

**Comment by charliermarsh** [comment]
One thing I'm realizing, just to confirm my understanding, is that in many cases, these will actually be fixed yet again by the f-string rule, right?

E.g. `'%s %s' % (a, b)` to `'{} {}'.format(a, b)` to `f'{a} {b}'`.

**Comment by colin99d** [comment]
> One thing I'm realizing, just to confirm my understanding, is that in many cases, these will actually be fixed yet again by the f-string rule, right?
> 
> E.g. `'%s %s' % (a, b)` to `'{} {}'.format(a, b)` to `f'{a} {b}'`.

That is correct!

**Comment by colin99d** [comment]
> One thing I'm realizing, just to confirm my understanding, is that in many cases, these will actually be fixed yet again by the f-string rule, right?
> 
> E.g. `'%s %s' % (a, b)` to `'{} {}'.format(a, b)` to `f'{a} {b}'`.

Also, the f-strings purposely ignores strings with format specifiers, so that UP030 can handle it first.

**Comment by colin99d** [comment]
@charliermarsh, I began working on the conversion to CFormatQuantity. This definitely looks like it will be a better way to do things! So far I have one question, how do I convert CConversionFlags to a string? So far I have been digging in [here](https://docs.rs/rustpython-common/0.2.0/rustpython_common/cformat/struct.CConversionFlags.html), and I have not been able to come up with anything.

**Comment by charliermarsh** [comment]
Hmm, I think I'd look at the [RustPython source](https://github.com/RustPython/RustPython/blob/59b191c789c891a500659d2c3b29a645883f780a/common/src/cformat.rs#L410) for those flags to understand how they're created, then implement the inverse? Does that work?

**Comment by colin99d** [comment]
> 

Oh perfect, thanks

**Comment by colin99d** [comment]
@charliermarsh, if you dont mind digging into this if you get some free time I think it might have some differences that are too big. For example, if the precision is just ".", this library decides to [make it 6](https://github.com/RustPython/RustPython/blob/59b191c789c891a500659d2c3b29a645883f780a/common/src/cformat.rs#L147). This is different to python, which would round to 0 decimals if precision is just 0. If we decide this doesn't work we can go back to the old way of doing it.

**Comment by charliermarsh** [comment]
(Will try to take a look tonight.)

**Comment by charliermarsh** [comment]
I've started poking at this but there's a lot to read :)

Are there any other discrepancies beyond what you've described? (Would it make sense to fix those in RustPython, if they're actual CPython incompatibilities?)

**Comment by colin99d** [comment]
> 

There are a couple, yeah. Also, I would be interested in knowing why they chose 6. We could attempt a PR there, or I could Just try to fix the one we have? Let me know what you want me to do.

**Comment by charliermarsh** [comment]
I think it's because of this?

```py
>>> "%f" % 2.12
'2.120000'
```

**Comment by colin99d** [comment]
You are right, but the issue is they are treating the below code as a 6. This means if we get a 6 we dont know if it should be a 0 or a 6.
Screenshot 2023-01-19 at 7 21 48 PM

**Comment by charliermarsh** [comment]
Yeah that, I think, should be fixed. I'll look into fixing it as I review this. Are there other limitations though?

**Comment by colin99d** [comment]
It looks like this and pyupgrade are disagreeing on another one. But I am not sure who is right. The pyupgrade tests is [here](https://github.com/asottile/pyupgrade/blob/main/tests/features/percent_format_test.py#L23). Pyupgrade wants to convert this into a statement, but rustpython doesn't. It looks like rustpython might be right though:
```
>>> "%%" % 2.12
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: not all arguments converted during string formatting
>>> "%%" % "hello"
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: not all arguments converted during string formatting
>>>
```

**Comment by charliermarsh** [comment]
Yeah that seems ok. So if I go and fix the RustPython parse limitation, does this seem good to go from your perspective?

**Comment by colin99d** [comment]
I am getting a little more weird behavior, but for right now im guessing its on me. I will let you know by EOD today if this is the only issue.

**Comment by charliermarsh** [comment]
I guess a related issue is that with:

```
"%f" % 1
```

I think RustPython reports that as precision(6), so I'm not sure we can reconstruct it exactly?

**Comment by colin99d** [comment]
> 

Yes this issue is definitely on them, but there is some other stuff im seeing that is weird. This is good though. We get to help the whole rust/python ecosystem on this stuff.

**Comment by charliermarsh** [comment]
I'll put together a PR to fix those limitations in RustPython now.

**Comment by charliermarsh** [comment]
To be clear, though, I am impressed by your regex abilities!

**Comment by colin99d** [comment]
> To be clear, though, I am impressed by your regex abilities!

Haha, 2 weeks ago I would not have known that. Regex101, and a buddy I know who did them full time for a couple months got me a lot further along than I should be (and some are just copied from pyupgrade).

**Comment by colin99d** [comment]
> Yeah, my preference is to use the parser if we can. I think it will be faster and more robust (and easier for us to maintain, reason about, etc.) than regular expressions. So I want to push us down that path unless we can find a reason that it's ~impossible.

Yes. I would much rather rewrite now than deal with 100 edge case bugs once the feature gets merged.

**Comment by charliermarsh** [comment]
https://github.com/RustPython/RustPython/pull/4463

**Comment by charliermarsh** [comment]
BTW, I think Pyupgrade handles `%.f` incorrectly, maybe?

This code:

```py
x = 1
y = 2
z = 3

print('%.2f %.f %f' % (x, y, z))
```

Gets changed to:

```py
x = 1
y = 2
z = 3

print('{:.2f} {:.f} {:f}'.format(x, y, z))
```

Which yields:

```
Traceback (most recent call last):
  File "/Users/crmarsh/workspace/ruff/foo.py", line 5, in <module>
    print('{:.2f} {:.f} {:f}'.format(x, y, z))
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
ValueError: Format specifier missing precision
```

**Comment by colin99d** [comment]
> 

Looks like you are right, maybe we keep the 6 as a default if they do not specify, OR we could just issue a warning but not fix. I like the first option a little more, but as always up to you.

**Comment by charliermarsh** [comment]
@colin99d - Mind if I do a pass on this from the current state of the code?

**Comment by colin99d** [comment]
> @colin99d - Mind if I do a pass on this from the current state of the code?

I have some bad code im fixing. Unit tests caught a new edge case. Let me finish and then she's all yours.

**Comment by charliermarsh** [comment]
Okay thanks. Just LMK when it's ready. I'll point to the updated RustPython branch and get it into a finished state, hopefully.

**Comment by charliermarsh** [comment]
And yeah -- I think we map `%.f` to `%.0f`. Same with `%.s` etc.

**Comment by colin99d** [comment]
@charliermarsh I am ready for you. Running `cargo test printf_string_formatting` currently results in two failures. Both are related to the precision issue.

**Comment by charliermarsh** [comment]
Working my way through this one, found at least one bug but I think it might be in our logic and not the parser.

**Comment by charliermarsh** [comment]
@colin99d - Ok, I think this is passing all tests now. I can think of a few ways that it could be improved so I might iterate on it a bit more. (For example: it'd be nice if we could avoid transforms that create really long string format calls.) It's a really significant transform, so I want it to be rock-solid before it goes out.

**Comment by colin99d** [comment]
> 

One thing I did in another check, is that if the new string is longer I would not do the change.

**Comment by charliermarsh** [comment]
I guess I'm also thinking about whether we should be fixing multi-line strings.

**Comment by colin99d** [comment]
I can see it either way. Pyupgrade did NOT fix the strings that you sent me earlier that broke.

**Comment by charliermarsh** [comment]
Alright, I'm gonna merge this as soon as #4463 is released.
